gubernator.go

```diff
@@ -305,74 +308,74 @@ func (s *Instance) getRateLimit(r *RateLimitReq) (*RateLimitResp, error) {
 	return nil, errors.Errorf("invalid rate limit algorithm '%d'", r.Algorithm)
 }
 
-// SetPeers is called when the pool of peers changes
-func (s *Instance) SetPeers(peers []PeerInfo) {
-	picker := s.conf.Picker.New()
-	var errs []string
-
-	for _, peer := range peers {
-		peerInfo := s.conf.Picker.GetPeerByHost(peer.Address)
-		// If we don't have an existing connection, we need to open one.
-		if peerInfo == nil {
-			var err error
-			peerInfo, err = NewPeerClient(s.conf.Behaviors, peer.Address)
-			if err != nil {
-				errs = append(errs,
-					fmt.Sprintf("failed to connect to peer '%s'; consistent hash is incomplete", peer.Address))
-				continue
+// SetPeers is called by the implementor to indicate the pool of peers has changed
+func (s *Instance) SetPeers(peerInfo []PeerInfo) {
+	localPicker := s.conf.LocalPicker.New()
+	regionPicker := s.conf.RegionPicker.New()
+
+	for _, info := range peerInfo {
+		// Add peers that are not in our local DC to the RegionPicker
+		if info.DataCenter != s.conf.DataCenter {
+			peer := s.conf.RegionPicker.GetByPeerInfo(info)
+			// If we don't have an existing PeerClient create a new one
+			if peer == nil {
+				peer = NewPeerClient(s.conf.Behaviors, info)
 			}
-
+			regionPicker.Add(peer)
+			continue
 		}
-
-		// If this peer refers to this server instance
-		peerInfo.isOwner = peer.IsOwner
-
-		picker.Add(peerInfo)
+		// If we don't have an existing PeerClient create a new one
+		peer := s.conf.LocalPicker.GetByPeerInfo(info)
+		if peer == nil {
+			peer = NewPeerClient(s.conf.Behaviors, info)
+		}
+		localPicker.Add(peer)
 	}
 
 	s.peerMutex.Lock()
+	// Replace our current pickers
+	oldLocalPicker := s.conf.LocalPicker
+	s.conf.LocalPicker = localPicker
+	s.conf.RegionPicker = regionPicker
+	s.peerMutex.Unlock()
 
-	oldPicker := s.conf.Picker
-
-	// Replace our current picker
-	s.conf.Picker = picker
-
-	// Update our health status
-	s.health.Status = Healthy
+	// TODO: We should run this in a routine and query the PeerPickers for their last errors, If any errors then
+	//  cluster is unhealthy. When our node is talking to other nodes, we should record all their errors for a time.
+	//  Some type of expire cache I think, where errors will eventually expire, but if someone hits our `/healthy`
+	//  endpoint we should return un-healthy if there are errors in the error cache
```

Comment:
Should not be handled in this PR.

Reply:
okay, I lied about this. Since we are planning on implementing multi-region support in 2 PR's then we shouldn't merge this PR to master with the `/healthcheck` endpoint broken. Let me know if you need help working on this part. Should be realtively easy to implement.